subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyMetadataRulesIntegrationTest.groovy

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.integtests.resolve
+
+import org.gradle.integtests.fixtures.AbstractHttpDependencyResolutionTest
+import org.gradle.integtests.fixtures.resolve.ResolveTestFixture
+import org.gradle.test.fixtures.HttpRepository
+import org.gradle.test.fixtures.Module
+import org.gradle.test.fixtures.maven.MavenFileRepository
+import spock.lang.Unroll
+
+abstract class DependencyMetadataRulesIntegrationTest extends AbstractHttpDependencyResolutionTest {
```

Comment:
❌ Missing test case: Resolving one configuration does not influence the result of resolving another configuration. Or in other words: Make sure that metadata rules don't accidentally modify the immutable metadata in the cache.

I'd test this by resolving two configurations that have the same dependencies and having a rule which 

1. Checks that some transitive dependency is in the metadata
2. Removes that transitive

The first check would fail when the rule is run the second time and if the cached metadata had been accidentally adjusted.

Reply:
Good idea. I added the test case.